libs/posthog_client_lite/src/lib.rs

```diff
@@ -0,0 +1,433 @@
+                    return None;
+                };
+                let Ok(requested) = requested.parse::<f64>() else {
+                    return None;
+                };
+                // Left can either be a number or a string
+                let provided = match provided {
+                    PostHogFlagFilterPropertyValue::Number(provided) => *provided,
+                    PostHogFlagFilterPropertyValue::String(provided) => {
+                        let Ok(provided) = provided.parse::<f64>() else {
+                            return None;
+                        };
+                        provided
+                    }
+                    _ => return None,
+                };
+                match operator {
+                    "lt" => Some(provided < requested),
+                    "gt" => Some(provided > requested),
+                    _ => None,
+                }
+            }
+            _ => None,
+        }
+    }
+
+    /// Evaluate a percentage.
+    fn evaluate_percentage(&self, mapped_user_id: f64, percentage: f64) -> bool {
+        mapped_user_id <= percentage
+    }
+
+    /// Evaluate a filter group for a feature flag. Returns `None` if the group is not matched or if there are errors
+    /// during the evaluation.
+    fn evaluate_group(
+        &self,
+        group: &PostHogLocalEvaluationFlagFilterGroup,
+        mapped_user_id: f64,
+        provided_properties: &HashMap<String, PostHogFlagFilterPropertyValue>,
+    ) -> Option<String> {
+        if let Some(ref properties) = group.properties {
+            for property in properties {
+                if let Some(value) = provided_properties.get(&property.key) {
+                    // The user provided the property value
+                    if self.evaluate_condition(property.operator.as_ref(), value, &property.value)
+                        != Some(true)
+                    {
+                        return None;
+                    }
+                } else {
+                    // We cannot evaluate, the property is not available
+                    return None;
+                }
+            }
+            // If all conditions are met, return the variant
+            if self.evaluate_percentage(mapped_user_id, group.rollout_percentage / 100.0) {
+                group.variant.clone()
+            } else {
+                None
+            }
+        } else {
+            // No matchers, apply to all users
+            if self.evaluate_percentage(mapped_user_id, group.rollout_percentage / 100.0) {
+                group.variant.clone()
+            } else {
+                None
+            }
+        }
+    }
+
+    /// Evaluate a multivariate feature flag. Returns `None` if the flag is not available or if there are errors
+    /// during the evaluation.
+    ///
+    /// The parsing logic is as follows:
+    ///
+    /// * Match each filter group. If any group is matched, check the rollout percentage against the precomputed
```

Comment:
Is this logic documented somewhere in PostHog? Or is it our add-hoc logic?

It looks a little bit strange. E.g. if the user matches 2 groups, each has 50% chance. The user will be chosen with 75% chance (1 - 0.5 * 0.5). Is it intended?

And if a user matches a group with 0% chance and wasn't selected, we evaluate them by the global rollout percentage anyway. So it's not possible to disable a feature for a user/group. Is it intended?

Reply:
This is (almost) the same logic as PostHog's Python client.

https://github.com/PostHog/posthog-python/blob/7aea6b72d3c74ffb580d65f86437704a90f6eb5f/posthog/feature_flags.py#L69-L81

> And if a user matches a group with 0% chance and wasn't selected, we evaluate them by the global rollout percentage anyway. So it's not possible to disable a feature for a user/group. Is it intended?

To disable a feature to a specific customer, we should set a condition tenant_id = X, rollout to 100%, variant = disabled?

Reply:
...and the order of conditions matters: that rule should appear before the rollout percentage = 0 condition.

Reply:
I reread the PostHog Python client and updated some behavior of the code -- they are now documented in the rustdoc.